### PR TITLE
Address time zone localization issue

### PR DESCRIPTION
**Pull request by john-bodley**
## Description

This PR addresses https://github.com/trinodb/trino-python-client/issues/366. 

Per [this](https://blog.ganssle.io/articles/2018/03/pytz-fastest-footgun.html) article, there seemed to be merit in removing `pytz` in favor of `dateutil.tz` given that in Python 3.6 the `datetime` module remedied the ambiguous datetimes due to daylight saving time transition. 

Regrettably though one cannot obtain the IANA name from a `dateutil.tz` time zone—which is required when transpiling `datetime.time`/`datetime.datetime` objects to Trino SQL when provided as operation parameters. Instead we use the `zoneinfo` package (added to Python’s standard library in Python 3.9 and back ported) which does provide a mechanism of obtaining the specified IANA name from a `datetime.datetime`/`datetime.time` object, i.e., 

```python
>>> from datetime import time
>>> from zoneinfo import ZoneInfo
>>>
>>> midnight = time(0, tzinfo=ZoneInfo("America/Los_Angeles"))
>>> midnight.tzinfo.key 
'America/Los_Angeles'
```

The fix is rather simple as simply switching out the `pytz` package for the `zoneinfo` package remedies the issue:

```
>>> from datetime import datetime
>>> import pytz
>>>
>>> datetime(2023, 1, 1, tzinfo=pytz.timezone("America/Los_Angeles")).isoformat()
'2023-01-01T00:00:00-07:53' 
```

```
>>> from datetime import datetime
>>> import zoneinfo
>>>
>>> datetime(2023, 1, 1, tzinfo=zoneinfo.ZoneInfo("America/Los_Angeles")).isoformat()
'2023-01-01T00:00:00-08:00'
```

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
* Fixes time zone localization. The `pytz` package for encoding/decoding time zones has been deprecated in favor of `zoneinfo`. Time zones should now be encoded (when specifying execution parameters) using the `zoneinfo.ZoneInfo` method as opposed to `pytz.timezone` method. ({[issue](https://github.com/trinodb/trino-python-client/issues/366.)}`366`)
```

### CODE REVIEW

File: tests/integration/test_types_integration.py [hunk 1]
```diff
@@ -284,13 +288,15 @@ def test_time(trino_connection):
...
 
 @pytest.mark.skipif(trino_version() == '351', reason="time not rounded correctly in older Trino versions")
-def test_time_with_timezone(trino_connection):
-    query_time_with_timezone(trino_connection, '-08:00')
-    query_time_with_timezone(trino_connection, '+08:00')
-    query_time_with_timezone(trino_connection, '+05:30')
-
-
-def query_time_with_timezone(trino_connection, tz_str: str):
+@pytest.mark.parametrize(
```
Comment (john-bodley) [inline tests/integration/test_types_integration.py:291]: Same logic as before but more aligned with how `pytest` was intended to be used.

File: trino/client.py [hunk 2]
```diff
@@ -51,21 +51,18 @@
...
 from tzlocal import get_localzone_name  # type: ignore
 
 import trino.logging
 from trino import constants, exceptions
 
-try:
-    from zoneinfo import ZoneInfo  # type: ignore
-
-except ModuleNotFoundError:
-    from backports.zoneinfo import ZoneInfo  # type: ignore
```
Comment (john-bodley) [inline trino/client.py:66]: Previously there was no requirement in `setup.py` for the `backports.zoneinfo` package.

File: tests/integration/test_types_integration.py [hunk 3]
```diff
@@ -300,114 +306,114 @@ def query_time_with_timezone(trino_connection, tz_str: str):
         # min supported time(3)
         .add_field(
             sql=f"TIME '00:00:00 {tz_str}'",
-            python=time(0, 0, 0).replace(tzinfo=tz))
+            python=time(0, 0, 0, tzinfo=tz))
```
Comment (john-bodley) [inline tests/integration/test_types_integration.py:309]: Clearer/cleaner to specify the time zone when creating the object.

**Review by mdesmet — APPROVED** [review]
LGTM

**Comment by john-bodley** [comment]
@mdesmet I've addressed your comments.

**Comment by hashhar** [comment]
I'll take a look tomorrow. I'm leaving some quick editorial comments for now.

Since this is a correctness issue I'd like to do a release as soon as we have this merged.

**Review by hashhar — COMMENTED** [review]
Commit "[preparatory] Use @pytest.mark.parameterize".

The project avoids prefixing commit messages, so something just like `Parameterize time and timestamp with time zone tests` is good enough.

File: tests/integration/test_types_integration.py (hunk 3 above)
Comment (hashhar) [reply tests/integration/test_types_integration.py:309]: are there any other occurrences of this in production code as well?

**Review by hashhar — COMMENTED** [review]
commit "[preparatory] Specify tzinfo during datetime/time object creation" -> "Specify tzinfo during datetime/time creation in tests"

File: setup.py [hunk 4]
```diff
@@ -78,7 +78,12 @@
         "Topic :: Database :: Front-Ends",
     ],
     python_requires='>=3.7',
-    install_requires=["pytz", "requests", "tzlocal"],
+    install_requires=[
+        "backports.zoneinfo;python_version<'3.9'",
```
Comment (hashhar) [inline setup.py:82]: missing zoneinfo require should be a separate commit, it's not related to the fix itself.

File: trino/client.py [hunk 5]
```diff
@@ -1041,7 +1037,7 @@ def new_instance(self, value: datetime, fraction: Decimal) -> TimestampWithTimeZ
         return TimestampWithTimeZone(value, fraction)
 
     def normalize(self, value: datetime) -> datetime:
-        if isinstance(self._whole_python_temporal_value.tzinfo, BaseTzInfo):
+        if tz.datetime_ambiguous(value):
```
Comment (hashhar) [inline trino/client.py:1040]: nice.

**Review by hashhar — COMMENTED** [review]
commit " [fix] Address timezone localization issue ".

Can you add a short explanation in the commit message body about the issue? Alternatively I think it'd also be self-explanatory if you can add a test which showcases the issue and then in next commit add the fix and adjust the test.

**Comment by john-bodley** [comment]
@hashhar I've addressed your comments.

**Review by hashhar — APPROVED** [review]
